In [14]:
import pandas as pd
from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA
from sklearn.cluster import KMeans
import hvplot.pandas

In [15]:
# Loading processed iris data 
file_path = './Resources/new_iris_data.csv'
df_iris = pd.read_csv(file_path)
df_iris.head()

sepal_length  petal_length  sepal_width  petal_width
0           5.1           1.4          3.5          0.2
1           4.9           1.4          3.0          0.2
2           4.7           1.3          3.2          0.2
3           4.6           1.5          3.1          0.2
4           5.0           1.4          3.6          0.2

In [16]:
# Standardize data with StandardScaler
iris_scaled = StandardScaler().fit_transform(df_iris)
print(iris_scaled[0:5])

[[-0.90068117 -1.3412724   1.03205722 -1.31297673]
 [-1.14301691 -1.3412724  -0.1249576  -1.31297673]
 [-1.38535265 -1.39813811  0.33784833 -1.31297673]
 [-1.50652052 -1.2844067   0.10644536 -1.31297673]
 [-1.02184904 -1.3412724   1.26346019 -1.31297673]]


In [17]:
# Initialize PCA model
pca = PCA(n_components=2)

In [18]:
# Get two principal components for the iris data.
iris_pca = pca.fit_transform(iris_scaled)

In [19]:
# Tranform PCA data into Dataframe
df_iris_pca = pd.DataFrame(
    data = iris_pca, columns = ['Princple Component 1', 'Principle Component 2'])
df_iris_pca.head()

Princple Component 1  Principle Component 2
0             -2.264542               0.505704
1             -2.086426              -0.655405
2             -2.367950              -0.318477
3             -2.304197              -0.575368
4             -2.388777               0.674767

In [20]:
# Fetch the explained variance 
pca.explained_variance_ratio_

array([0.72770452, 0.23030523])

What this tells us, is that the first principal component contains 72.77% of the variance and the second contains 23.03%. Together, they contain 95.80% of the information.

In [21]:
# Find the best value for K
inertia = []
k = list(range(1, 11))

# Calculate the inertia for the range of K values
for i in k:
    km = KMeans(n_clusters=i, random_state=0)
    km.fit(df_iris_pca)
    inertia.append(km.inertia_)

# Create the elbow curve
elbow_data = {"k": k, "inertia": inertia}
df_elbow = pd.DataFrame(elbow_data)
df_elbow.hvplot.line(x="k", y="inertia", xticks=k, title="Elbow Curve")

C:\Users\snkty\anaconda3\envs\mlenv\lib\site-packages\sklearn\cluster\_kmeans.py:882: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  f"KMeans is known to have a memory leak on Windows "


:Curve   [k]   (inertia)

In [22]:
# Initialize the K-means model
model = KMeans(n_clusters=3, random_state=0)

# Fit the model
model.fit(df_iris_pca)

# Predict clusters
predictions = model.predict(df_iris_pca)

# Add the predicted class columns
df_iris_pca["class"] = model.labels_
df_iris_pca.head()

Princple Component 1  Principle Component 2  class
0             -2.264542               0.505704      0
1             -2.086426              -0.655405      0
2             -2.367950              -0.318477      0
3             -2.304197              -0.575368      0
4             -2.388777               0.674767      0

In [23]:
df_iris_pca.hvplot.scatter(
    x="Princple Component 1",
    y="Principle Component 2",
    hover_cols=["class"],
    by="class",
)

:NdOverlay   [class]
   :Scatter   [Princple Component 1]   (Principle Component 2,class)